src/server.rs

```diff
@@ -227,855 +725,233 @@ impl ServerStats {
+        }
     }
 
-    /// Return a clone of the object implementing `CommandCreatorSync` that this server uses to create processes.
-    ///
-    /// This is intended for use in testing. In non-testing, this will
-    /// just return a `ProcessCommandCreator` which is a unit struct.
-    #[allow(dead_code)]
-    pub fn command_creator(&self) -> C {
-        self.creator.clone()
-    }
-
-    /// Register Server with the event loop.
-    fn register(&mut self, mut event_loop: &mut EventLoop<SccacheServer<C>>) -> io::Result<()> {
-        event_loop.register(
-            &self.sock,
-            self.token,
-            EventSet::readable(),
-            PollOpt::edge() | PollOpt::oneshot()
-        ).and_then(|_| {
-            self.reset_idle_timer(&mut event_loop);
-            Ok(())
-        })
+    fn poll_complete(&mut self) -> Poll<(), io::Error> {
+        self.inner.poll_complete()
     }
+}
 
-    /// Re-register Server with the event loop.
-    fn reregister(&mut self, event_loop: &mut EventLoop<SccacheServer<C>>) {
-        if !self.shutting_down {
-            event_loop.reregister(
-                &self.sock,
-                self.token,
-                EventSet::readable(),
-                PollOpt::edge() | PollOpt::oneshot()
-             ).unwrap_or_else(|_e| {
-                 let server_token = self.token;
-                 self.reset_connection(event_loop, server_token);
-             })
-        }
-    }
-
-    /// Accept a new client connection.
-    fn accept(&mut self, event_loop: &mut EventLoop<SccacheServer<C>>) {
-        debug!("Client connecting");
-        let sock = match self.sock.accept() {
-            Ok(Some((sock, _addr))) => sock,
-            Err(_) | Ok(None) => {
-                self.reregister(event_loop);
-                return;
-            }
-        };
-
-        if let Some(token) = self.conns.insert_with(|token|
-        {
-            ClientConnection::new(sock, token)
-        })
-        {
-            match self.conns[token].register(event_loop) {
-                Ok(_) => {},
-                Err(_e) => {
-                    self.conns.remove(token);
-                }
-            }
-        };
-
-
-        self.reregister(event_loop);
-    }
+impl<I: Io + 'static> Transport for SccacheTransport<I> {}
 
-    /// Reset a connection, either the listen socket or a client socket.
-    fn reset_connection(&mut self, event_loop: &mut EventLoop<SccacheServer<C>>, token: Token) {
-        if self.token == token {
-            if !self.shutting_down {
-                // Not actively trying to shut down, but something bad happened.
-                event_loop.shutdown();
-            }
-        } else {
-            debug!("reset connection; token={:?}", token);
-            self.conns.remove(token);
-            self.check_shutdown(event_loop);
-        }
-    }
+/// Simple tokio-core `Codec` which uses stock protobuf functions to
+/// decode/encode protobuf messages.
```

Comment:
I love this part of tokio's design! When I eventually rip out protobuf and replace it with something based on serde this will make it a lot easier.